File: oneflow/python/nn/consistent_cast.py
"""
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
"""
from __future__ import absolute_import
import re
import functools
from typing import Union, Optional, Tuple, List, Callable
import oneflow as flow
from oneflow.python.oneflow_export import oneflow_export
from oneflow.python.nn.module import Module
import oneflow.python.framework.session_context as session_ctx
import oneflow.python.nn.consistent_cast_util as consistent_cast_util


@oneflow_export("consistent_cast")
def api_consistent_cast(
    mirrored_entity: Callable,
    parallel_distribution: Tuple[
        List[Union[Tuple[str], str]], List[Union[Tuple[str], str]]
    ],
    placement_signature: Optional[
        Tuple[List[flow.placement], List[flow.placement],]
    ] = None,
):
    assert callable(mirrored_entity), "mirrored_entity is not callable"

    def check_input_is_valid(parallel_distribution, placement_signature):
        assert len(parallel_distribution) == len(placement_signature)
        for i in range(len(parallel_distribution)):
            if isinstance(parallel_distribution[i], (tuple, list)):
                assert len(parallel_distribution[i]) == len(
                    placement_signature[i].hierarchy
                )
                for sbp_parallel_str in parallel_distribution:
                    assert re.match("^B|P|(S\(\d+\))$", sbp_parallel_str) is not None, (
                        "Distribute %s is not valid" % sbp_parallel_str
                    )
            else:
                assert (
                    re.match("^B|P|(S\(\d+\))$", parallel_distribution[0]) is not None
                ), ("Distribute %s is not valid" % parallel_distribution)
                assert len(placement_signature[i].hierarchy) == 1

    if placement_signature is None:
        cur_scope = flow.current_scope()
        placement_signature = (
            [cur_scope.device_parallel_desc_symbol for _ in parallel_distribution[0]],
            [cur_scope.device_parallel_desc_symbol for _ in parallel_distribution[1]],
        )
    check_input_is_valid(parallel_distribution[0], placement_signature[0])
    check_input_is_valid(parallel_distribution[1], placement_signature[1])

    if isinstance(mirrored_entity, Module):
        return module_consistent_cast(
            mirrored_entity, parallel_distribution, placement_signature
        )
    else:
        return function_consistent_cast(
            mirrored_entity, parallel_distribution, placement_signature
        )


def module_consistent_cast(
    module: Module,
    parallel_distribution: Tuple[
        List[Union[Tuple[str], str]], List[Union[Tuple[str], str]]
    ],
    placement_signature: Optional[
        Tuple[List[flow.placement], List[flow.placement],]
    ] = None,
):
    assert (
        not module.consistent
    ), "the module is already consistented module, don't cast again!"
    return module.consistent_cast(parallel_distribution, placement_signature)


def function_consistent_cast(
    mirrored_func: Callable,
    parallel_distribution: Tuple[
        List[Union[Tuple[str], str]], List[Union[Tuple[str], str]]
    ],
    placement_signature: Optional[
        Tuple[List[flow.placement], List[flow.placement],]
    ] = None,
):
    @functools.wraps(mirrored_func)
    def wrapped_func(*args):
        args = list(args)
        sess = session_ctx.GetDefaultSession()
        with sess.consistent_scope():
            args = consistent_cast_util.cast_input_to_consistent(
                args, parallel_distribution[0], placement_signature[0]
            )
            consistent_output = mirrored_func(*args)
            return consistent_cast_util.cast_output_from_consistent(
                consistent_output, parallel_distribution[1], placement_signature[1]
            )

    return wrapped_func

File: oneflow/core/autograd/gradient_funcs/consistent_cast.cpp
/*
Copyright 2020 The OneFlow Authors. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
#include "oneflow/core/framework/op_expr_grad_function.h"
#include "oneflow/core/framework/op_interpreter/op_interpreter_util.h"
#include "oneflow/core/framework/nd_sbp.h"
#include "oneflow/core/boxing/eager_boxing_interpreter_mgr.h"
#include "oneflow/core/framework/tensor_rpc_util.h"
#include "oneflow/core/common/decorator.h"
#include "oneflow/core/functional/functional.h"

namespace oneflow {
namespace one {

struct CastConsistentCaptureState : public AutoGradCaptureState {
  Symbol<ParallelDesc> parallel_desc;
  Symbol<NdSbp> nd_sbp;
  std::shared_ptr<const Shape> shape;
  Symbol<DType> dtype;
};

class CastToConsistent : public OpExprGradFunction<CastConsistentCaptureState> {
 public:
  Maybe<void> Init(const OpExpr& op) override {
    const auto* fw_op_expr = dynamic_cast<const CastToConsistentOpExpr*>(&op);
    CHECK_NOTNULL_OR_RETURN(fw_op_expr);  // NOLINT(maybe-need-error-msg)
    const std::string& op_name = fw_op_expr->op_name();
    grad_op_ = JUST(one::CastFromConsistentOpExpr::New(GradientOpName(op_name)));
    return Maybe<void>::Ok();
  }

  Maybe<void> Capture(CastConsistentCaptureState* ctx, const TensorTuple& inputs,
                      const TensorTuple& outputs,
                      const OpExprInterpContext& interp_ctx) const override {
    ctx->parallel_desc = JUST(interp_ctx.parallel_desc);
    ctx->nd_sbp = JUST(GetDualNdSbp(JUST(interp_ctx.nd_sbp)));
    return Maybe<void>::Ok();
  }

  Maybe<void> Apply(const CastConsistentCaptureState* ctx, const TensorTuple& out_grads,
                    TensorTuple* in_grads) const override {
    CHECK_EQ_OR_RETURN(out_grads.size(), 1);  // NOLINT(maybe-need-error-msg)
    std::shared_ptr<Tensor> out_grad = out_grads.at(0);
    CHECK_OR_RETURN(out_grad->is_consistent())
        << Error::RuntimeError()
        << "Expected global tensor for cast_to_consistent but got local tensor";
    {
      Symbol<NdSbp> nd_sbp_constraint = ctx->nd_sbp;
      Symbol<ParallelDesc> parallel_desc_constraint = ctx->parallel_desc;
      out_grad = JUST(functional::ToConsistent(out_grad, parallel_desc_constraint,
                                               *JUST(GetSbpList(nd_sbp_constraint)),
                                               GetNoneSbpList(), /* check_meta */ false));
    }
    in_grads->at(0) = JUST(OpInterpUtil::Dispatch<Tensor>(*grad_op_, {out_grad}));
    return Maybe<void>::Ok();
  }

 private:
  std::shared_ptr<OpExpr> grad_op_;
};

REGISTER_OP_EXPR_GRAD_FUNCTION("cast_to_consistent", CastToConsistent);

class CastFromConsistent : public OpExprGradFunction<CastConsistentCaptureState> {
 public:
  Maybe<void> Init(const OpExpr& op) override {
    const auto* fw_op_expr = dynamic_cast<const CastFromConsistentOpExpr*>(&op);
    CHECK_NOTNULL_OR_RETURN(fw_op_expr);  // NOLINT(maybe-need-error-msg)
    const std::string& op_name = fw_op_expr->op_name();
    grad_op_ = JUST(one::CastToConsistentOpExpr::New(GradientOpName(op_name)));
    return Maybe<void>::Ok();
  }

  Maybe<void> Capture(CastConsistentCaptureState* ctx, const TensorTuple& inputs,
                      const TensorTuple& outputs, const AttrMap& attrs) const override {
    const auto& input = inputs.at(0);
    CHECK_OR_RETURN(input->is_consistent())
        << Error::RuntimeError()
        << "Expected global tensor for cast_from_consistent but got local tensor";
    ctx->parallel_desc = JUST(input->parallel_desc());
    ctx->nd_sbp = JUST(input->nd_sbp());
    ctx->shape = input->shape();
    ctx->dtype = input->dtype();
    return Maybe<void>::Ok();
  }

  Maybe<void> Apply(const CastConsistentCaptureState* ctx, const TensorTuple& out_grads,
                    TensorTuple* in_grads) const override {
    const auto& dual_nd_sbp = JUST(GetDualNdSbp(ctx->nd_sbp));
    MutableAttrMap attrs;
    JUST(attrs.SetAttr<Shape>("shape", *ctx->shape));
    JUST(attrs.SetAttr<DataType>("dtype", ctx->dtype->data_type()));
    in_grads->at(0) = JUST(OpInterpUtil::Dispatch<Tensor>(
        *grad_op_, {out_grads.at(0)}, OpExprInterpContext(attrs, ctx->parallel_desc, dual_nd_sbp)));
    return Maybe<void>::Ok();
  }

 private:
  std::shared_ptr<OpExpr> grad_op_;
};

REGISTER_OP_EXPR_GRAD_FUNCTION("cast_from_consistent", CastFromConsistent);

}  // namespace one
}  // namespace oneflow
